### PR TITLE
probes: add IPC4 support

### PR DESCRIPTION
**Pull request by abonislawski**
Add IPC4 support for extraction & injection probes
IPC4 header changes, checksum calculation
moved checksum from header to the end of data packet
TODO: still need a little cleanup in headers + get large config + delete PROBE_PURPOSE_LOGGING/TRACING and use module id 0 for init

### CODE REVIEW

**Comment by abonislawski** [comment]
Required patch for injection probes: #6199

File: src/include/sof/probe/probe.h [hunk 1]
```diff
@@ -74,12 +66,12 @@ int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size);
...
- * \brief Set probe points
+ * \brief Get info about connected injection DMAs
  *
- * param[in] count - number of probe points configured this call
- * param[in] probe - array of size 'count' with configuration of probe points
+ * param[in,out] data - reply to write data to
+ * param[in] max_size - maximum number of bytes available in data
  */
-int probe_point_add(uint32_t count, struct probe_point *probe);
+int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size);
```
Comment (lyakh) [inline src/include/sof/probe/probe.h:74]: why did you reorder these? They even were in alphabetic order.

File: src/probe/probe.c [hunk 2]
```diff
@@ -643,12 +626,20 @@ static int probe_gen_header(uint32_t buffer_id, uint32_t size,
...
+		    header->buffer_id  +
+		    header->format +
+		    header->timestamp_high +
+		    header->timestamp_low +
+		    header->data_size_bytes;
+#else
 	/* calc crc to check validation by probe parse app */
-	crc = crc32(0, header, sizeof(*header));
-	header->checksum = crc;
+	header->crc = 0;
```
Comment (lyakh) [inline src/probe/probe.c:640]: I don't think you renamed `.checksum` to `.crc` in this patch

File: src/probe/probe.c [hunk 3]
```diff
@@ -964,6 +965,42 @@ static bool probe_purpose_needs_ext_dma(uint32_t purpose)
...
 }
 
+#if CONFIG_IPC_MAJOR_4
+static struct comp_buffer *ipc4_get_buffer(struct ipc_comp_dev *dev, probe_point_id_t probe_point)
+{
+	struct comp_buffer *buf;
+	struct list_item *sink_list, *source_list;
+
+	if (probe_point.fields.type == PROBE_TYPE_INPUT) {
+		if (!list_is_empty(&dev->cd->bsource_list)) {
```
Comment (lyakh) [inline src/probe/probe.c:975]: I'd just do
```
if (list_is_empty(&dev->cd->bsource_list))
	return NULL;
```
here and below

File: src/probe/probe.c [hunk 4]
```diff
@@ -993,7 +1032,6 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 		    probe[i].purpose != PROBE_PURPOSE_LOGGING) {
 			tr_err(&pr_tr, "probe_point_add() error: invalid purpose %d",
 			       probe[i].purpose);
-
```
Comment (lyakh) [inline src/probe/probe.c:996]: unrelated

File: src/probe/probe.c [hunk 5]
```diff
@@ -1099,23 +1152,69 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
...
+		tr_err(&pr_tr, "probe_dma_info(): Not initialized.");
+
+		return -EINVAL;
+	}
+
+	data->rhdr.hdr.size = sizeof(*data);
+
+	/* search all injection DMAs to send them in reply */
+	while (i < CONFIG_PROBE_DMA_MAX &&
+	       data->rhdr.hdr.size + sizeof(struct probe_dma) < max_size) {
```
Comment (lyakh) [inline src/probe/probe.c:1198]: a `for` loop seems to fit better here

File: src/probe/probe.c [hunk 6]
```diff
@@ -1198,3 +1309,90 @@ int probe_point_remove(uint32_t count, uint32_t *buffer_id)
...
+{
+	comp_dbg(dev, "probe_set_large_config()");
+
+	switch (param_id) {
+	case IPC4_PROBE_MODULE_PROBE_POINTS_ADD:
+		return probe_point_add(data_offset / sizeof(struct probe_point), (struct probe_point *)data);
+	case IPC4_PROBE_MODULE_DISCONNECT_PROBE_POINTS:
+		return probe_point_remove(data_offset / sizeof(uint32_t), (uint32_t *)data);
+	case IPC4_PROBE_MODULE_INJECTION_DMA_ADD:
+		return probe_dma_add(data_offset / (2*sizeof(uint32_t)), (struct probe_dma *)data);
```
Comment (lyakh) [inline src/probe/probe.c:1359]: spaces around `*`

File: src/include/ipc4/probe.h [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+ */
+typedef union probe_point_id {
+	uint32_t full_id;
+	struct {
+		uint32_t  module_id   : 16;	/**< Target module ID */
+		uint32_t  instance_id : 8;	/**< Target module instance ID */
+		uint32_t  type        : 2;	/**< Probe point type as specified by ProbeType enumeration */
+		uint32_t  index       : 6;	/**< Queue index inside target module */
+	} fields;
+} probe_point_id_t;
```
Comment (kv2019i) [inline src/include/ipc4/probe.h:137]: Should this be packed as well?

File: src/probe/probe.c [hunk 8]
```diff
@@ -1099,23 +1152,69 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
...
 #endif
 		} else {
-			notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_PRODUCE,
+#if CONFIG_IPC_MAJOR_4
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_PRODUCE,
+					  &probe_cb_produce, 0);
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_FREE,
+					  &probe_cb_free, 0);
+#else
+			notifier_register(&probe[i].buffer_id.full_id, dev->cb, NOTIFIER_ID_BUFFER_PRODUCE,
```
Comment (kv2019i) [inline src/probe/probe.c:1168]: Hmm, this doesn't seem right, code is identical to IPC4 branch. "buffer_id.full_id" is in the IPC4 ABI only, right?

File: src/probe/probe.c [hunk 9]
```diff
@@ -1004,21 +1042,35 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 		}
 
 		if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION) {
-			/* check if buffer exists */
+#if CONFIG_IPC_MAJOR_4
+			dev = ipc_get_comp_by_id(ipc_get(), IPC4_COMP_ID(probe[i].buffer_id.fields.module_id,
```
Comment (kv2019i) [inline src/probe/probe.c:1046]: This is nice! I was worried we'd need more code to do the module based lookup, but it seems existing infra was pretty close already.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Looks really good @abonislawski  ! I few things in code that require work (or at least a double-check).

File: src/include/sof/probe/probe.h (hunk 1 above)
Comment (abonislawski) [reply src/include/sof/probe/probe.h]: Not needed, reverted

File: src/probe/probe.c (hunk 2 above)
Comment (abonislawski) [reply src/probe/probe.c:640]: Renamed in ipc3 header back to crc

File: src/probe/probe.c (hunk 5 above)
Comment (abonislawski) [reply src/probe/probe.c:1229]: old ipc3 code, just moved to another place

File: src/probe/probe.c (hunk 8 above)
Comment (abonislawski) [reply src/probe/probe.c:1199]: Yeap, that was my plan to unify this and use full_id everywhere so we dont have to branch for IPC all the time but forget to update ipc3 header which is done now :)

File: src/probe/probe.c (hunk 9 above)
Comment (abonislawski) [reply src/probe/probe.c:1077]: I was worried too! But with some extra for loops we can quickly find our buffer

**Review by lgirdwood — COMMENTED** [review]
LGTM, @kv2019i @jsarha @abonislawski I assume everything everything is ready in Zephyr and this has been tested with Linux driver PR ?

**Comment by abonislawski** [comment]
@lgirdwood tested with closed source validation tests, @jsarha will test with linux driver PR soon

File: src/probe/probe.c [hunk 10]
```diff
@@ -961,7 +962,72 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
...
+	struct comp_buffer *buf;
+	struct list_item *sink_list, *source_list;
+
+	if (probe_point.fields.type == PROBE_TYPE_INPUT) {
+		if (list_is_empty(&dev->cd->bsource_list))
+			return NULL;
+
+		list_for_item(source_list, &dev->cd->bsource_list) {
+			buf = container_of(source_list, struct comp_buffer, sink_list);
+
```
Comment (RanderWang) [inline src/probe/probe.c:984]: @lyakh   change ?
````c
struct comp_buffer __sparse_cache *buf_c;
buf_c = buffer_acquire(buf);
IPC4_SRC_QUEUE_ID(buf_c->id) ....
````

**Review by kv2019i — APPROVED** [review]
Looks good to me!

File: src/probe/probe.c [hunk 11]
```diff
@@ -1099,23 +1183,69 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
...
 #else
 			return -EINVAL;
 #endif
 		} else {
-			notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_PRODUCE,
+#if CONFIG_IPC_MAJOR_4
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_PRODUCE,
 					  &probe_cb_produce, 0);
-			notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_FREE,
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_FREE,
```
Comment (kv2019i) [inline src/probe/probe.c:1196]: Oops, one compiler warning, can you fix this before merge:

 1196 |                         notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_FREE,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1197 |                                           &probe_cb_free, 0);
      |                                           ~~~~~~~~~~~~~~~~~~

**Review by kv2019i — CHANGES_REQUESTED** [review]
One compiler warning that I missed, please check.

File: src/probe/probe.c (hunk 10 above)
Comment (lyakh) [reply src/probe/probe.c:984]: yes, good catch! On both `if` branches of course

File: src/probe/probe.c [hunk 12]
```diff
@@ -961,7 +962,72 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
...
+#endif
+}
+
+#if CONFIG_IPC_MAJOR_4
+static struct comp_buffer *ipc4_get_buffer(struct ipc_comp_dev *dev, probe_point_id_t probe_point)
+{
+	struct comp_buffer *buf;
+	struct list_item *sink_list, *source_list;
+
+	if (probe_point.fields.type == PROBE_TYPE_INPUT) {
```
Comment (lyakh) [inline src/probe/probe.c:978]: `switch (probe_point.fields.type)` would look better

File: src/probe/probe.c [hunk 13]
```diff
@@ -961,7 +962,72 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
...
+static bool enable_logs(struct probe_point *probe)
+{
+#if CONFIG_IPC_MAJOR_4
+	if (probe->buffer_id.fields.module_id == 0)
+#else
+	if (probe->purpose == PROBE_PURPOSE_LOGGING)
+#endif
+		return true;
+
+	return false;
```
Comment (lyakh) [inline src/probe/probe.c:1015]: please, make it
```
#if CONFIG_IPC_MAJOR_4
	return probe->buffer_id.fields.module_id == 0;
#else
	return probe->purpose == PROBE_PURPOSE_LOGGING;
#endif
```

File: src/probe/probe.c [hunk 14]
```diff
@@ -961,7 +962,72 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
...
+}
+
+static bool verify_purpose(uint32_t purpose)
+{
+#if CONFIG_IPC_MAJOR_4
+	if (purpose != PROBE_PURPOSE_EXTRACTION &&
+	    purpose != PROBE_PURPOSE_INJECTION)
```
Comment (lyakh) [inline src/probe/probe.c:1022]:
```
return purpose == PROBE_PURPOSE_EXTRACTION || purpose === PROBE_PURPOSE_INJECTION;
```

File: src/probe/probe.c [hunk 15]
```diff
@@ -961,7 +962,72 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
...
+#else
+	if (purpose != PROBE_PURPOSE_EXTRACTION &&
+	    purpose != PROBE_PURPOSE_INJECTION &&
+	    purpose != PROBE_PURPOSE_LOGGING)
```
Comment (lyakh) [inline src/probe/probe.c:1026]: similar

File: src/probe/probe.c [hunk 16]
```diff
@@ -985,12 +1054,12 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
...
 	for (i = 0; i < count; i++) {
 		tr_dbg(&pr_tr, "\tprobe[%u] buffer_id = %u, purpose = %u, stream_tag = %u",
-		       i, probe[i].buffer_id, probe[i].purpose,
+		       i, probe[i].buffer_id.full_id, probe[i].purpose,
 		       probe[i].stream_tag);
 
-		if (probe[i].purpose != PROBE_PURPOSE_EXTRACTION &&
-		    probe[i].purpose != PROBE_PURPOSE_INJECTION &&
-		    probe[i].purpose != PROBE_PURPOSE_LOGGING) {
+		fw_logs = enable_logs(&probe[i]);
```
Comment (lyakh) [inline src/probe/probe.c:1060]: maybe move this after the `if` below?

**Comment by lgirdwood** [comment]
@abonislawski I'm hearing good test results. Good to merge ? Some review comments/conflicts.

**Comment by abonislawski** [comment]
@lgirdwood yes, just need to apply review requests and fix CI

File: src/include/ipc/probe.h [hunk 17]
```diff
@@ -1,179 +1,10 @@
...
-struct sof_ipc_probe_point_remove_params {
-	struct sof_ipc_cmd_hdr hdr;	/**< Header */
-	uint32_t num_elems;		/**< Count of buffer IDs specified in array */
-	uint32_t buffer_id[];		/**< Array of buffer IDs from which Probe Points should be removed */
-} __attribute__((packed, aligned(4)));
 
-#endif /* __IPC_PROBE_H__ */
+#if CONFIG_IPC_MAJOR_4
+#include <ipc4/probe.h>
+#elif
```
Comment (jsarha) [inline src/include/ipc/probe.h:8]: This is wrong. #elfi requires an argument, just #else is enough.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kv2019i comments resolved ?

**Review by kv2019i — APPROVED** [review]

**Review by kv2019i — COMMENTED** [review]
@lgirdwood  @abonislawski  @jsarha Two things left, but we can also fix in follow-up PR as the probes are not enabled in builds by default:
- compiler warning on uninitialized "*buf" in add point (I marked this as accidentally resolved but it's still there)
- "undefined reference to `sys_comp_probe_init'" in IPC3 build if probes is enabled

**Comment by lgirdwood** [comment]
> * compiler warning on uninitialized "*buf" in add point (I marked this as accidentally resolved but it's still there)
> * "undefined reference to `sys_comp_probe_init'" in IPC3 build if probes is enabled

oh, ok - though that was resolved. @abonislawski can you fix and we can merge.

**Comment by abonislawski** [comment]
Sure
@kv2019i probes should be enabled by default for both IPC3 & IPC4?

**Comment by kv2019i** [comment]
@abonislawski wrote:
> Sure @kv2019i probes should be enabled by default for both IPC3 & IPC4?

That's probably something for another PR. Not sure why it has been disabled before. It would seem this should be enabled by default for al Intel platforms.

**Comment by abonislawski** [comment]
@kv2019i enabled by default for CAVS now and fixed some CI warnings, is it ok?

**Review by lgirdwood — APPROVED** [review]
